cylc/uiserver/data_store_mgr.py

```diff
@@ -0,0 +1,254 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Manage a local data-store replica of all workflow service data-stores.
+
+A local data-store is created and synced for all workflows established by
+the workflow service manager.
+
+The workflows publish the updated fields of the updated data elements (deltas),
+and these elements are grouped by type/topic. Once subscribed to, the publisher
+queues these messages until the are received, if the delta creation time is
+newer than that of the last update then it is applied (updates merged, pruned
+deleted) then a checksum is generated from the time stamped IDs and compared to
+the published one.
+
+Reconciliation on failed verification is done by requesting all elements of a
+topic, and replacing the respective data-store elements with this.
+
+Subscriptions are currently run in a different thread (via ThreadPoolExecutor).
+
+"""
+
+import asyncio
+import logging
+
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+from time import sleep
+
+from cylc.flow.network.server import PB_METHOD_MAP
+from cylc.flow.network.scan import MSG_TIMEOUT
+from cylc.flow.network.subscriber import WorkflowSubscriber, process_delta_msg
+from cylc.flow.data_store_mgr import (
+    EDGES, FAMILIES, FAMILY_PROXIES, JOBS, TASKS, TASK_PROXIES, WORKFLOW,
+    DELTAS_MAP, apply_delta, generate_checksum
+)
+from .workflows_mgr import workflow_request
+
+logger = logging.getLogger(__name__)
+
+
+class DataStoreMgr:
+    """Manage the local data-store acquisition/updates for all workflows."""
+
+    INIT_DATA_WAIT_TIME = 5.  # seconds
+    INIT_DATA_RETRY_DELAY = 0.5  # seconds
+    RECONCILE_TIMEOUT = 5.  # seconds
+
+    def __init__(self, workflows_mgr):
+        self.workflows_mgr = workflows_mgr
+        self.data = {}
+        self.w_subs = {}
+        self.topics = {topic.encode('utf-8') for topic in DELTAS_MAP}
+        self.topics.add(b'shutdown')
+        self.loop = None
+        # Might be options other than threads to achieve
+        # non-blocking subscriptions, but this works.
+        self.executor = ThreadPoolExecutor()
```

Comment:
I think this is probably the right way to go, as the subscriptions are presumably "I/O bound".

Reply:
Yeah, blocking.. However, there are some ways to avoid blocking.. i.e. Like not waiting around if there's no message(s) using the `flag=NOBLOCK` [recv](https://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Socket.recv) arg, or using a [poller](https://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Poller) .. but in that case you'd need to create a common list of sockets to loop over in a separate thread or as a timed callback in the main thread asyncio thread.. 
The current implementation is probably more event driven, and less burden/delay on the main thread and/or other subscriptions?

Open for other ideas.. But I think it should be a future improvement PR, if a better approach emerges.

Reply:
Not sure if we're on the same wavelength - I was just trying to say that threads is correct IMO, for IO-bound callables, as opposed to process pool - the other option with concurrent.futures. (Or asyncio? but yes, we can look later if needed...).

Reply:
Definitely on board.. I was just saying the blocking bit is receiving messages, but you can make it not hang-around/block while waiting for the next publish (don't block if no messages have been published). This could be an idea, even on a separate thread.

Reply:
Ah, roger that. :+1: